File: apps/ios/Sources/OlloCrypto/AccountVault.swift
import Foundation

/// Long-term account Ed25519 lives in its own wrapped slot. Generating a new
/// `AccountKey` on every unsigned launch would mint a new account each tap.
public final class AccountVault: @unchecked Sendable {
    private let store: IdentityStore
    private let wrapKey: Data

    public init(store: IdentityStore, wrapKey: Data) {
        self.store = store
        self.wrapKey = wrapKey
    }

    public func save(_ key: AccountKey) throws {
        try store.put(
            wrapKey: wrapKey,
            slot: .account,
            plaintext: try BlobMap.encode([
                "seed": key.privateKey.rawRepresentation,
                "public": key.publicKey,
            ])
        )
    }

    public func load() throws -> AccountKey? {
        guard let raw = try store.get(wrapKey: wrapKey, slot: .account) else { return nil }
        let map = try BlobMap.decode(raw)
        guard let seed = map["seed"] else { return nil }
        return try AccountKey(rawPrivateKey: seed)
    }

    public func getOrCreate() throws -> AccountKey {
        if let existing = try load() { return existing }
        let next = AccountKey()
        try save(next)
        return next
    }
}

File: apps/android/core/crypto/src/main/java/app/ollo/crypto/AccountVault.kt
package app.ollo.crypto

/**
 * Long-term account Ed25519 lives in its own wrapped slot. Generating a new
 * [AccountKey] on every unsigned launch would mint a new account each tap.
 */
class AccountVault(
    private val store: IdentityStore,
    private val wrapKey: ByteArray,
) {
    fun save(key: AccountKey) {
        store.put(
            wrapKey,
            IdentityStore.Slot.Account,
            BlobMap.encode(
                linkedMapOf(
                    "seed" to key.privateSeed.copyOf(),
                    "public" to key.publicKey.copyOf(),
                ),
            ),
        )
    }

    fun load(): AccountKey? {
        val raw = store.get(wrapKey, IdentityStore.Slot.Account) ?: return null
        val map = BlobMap.decode(raw)
        val seed = map["seed"] ?: return null
        val pub = map["public"] ?: return null
        return AccountKey.fromSeed(seed, pub)
    }

    fun getOrCreate(): AccountKey {
        load()?.let { return it }
        val next = AccountKey.generate()
        save(next)
        return next
    }
}
